### ISSUE TITLE
Computing target matrix differently

### ISSUE THREAD

**Issue by nishanthcgit** [openai/CLIP#31]
Now I realize that the released models and the model from the [keras code examples](https://keras.io/examples/nlp/nl_image_search/)  are different, and are possibly trained differently. As far as I can see, as per [this issue](https://github.com/openai/CLIP/issues/3) the openai CLIP model uses a target matrix of torch.eye(batch_size), while in the keras code examples:

> To calculate the loss, we compute the pairwise dot-product similarity between each caption_i and images_j in the batch as the predictions. The target similarity between caption_i and image_j is computed as the average of the (dot-product similarity between caption_i and caption_j) and (the dot-product similarity between image_i and image_j). Then, we use crossentropy to compute the loss between the targets and the predictions.

So the target in case of the image and caption not being matched isn't a 0 but the average of the distances between the image embeddings and text embeddings. I realize these are 2 different approaches to train the model but do you think setting the target to be the average instead of just a zero might help convergence/accuracy?

**Comment by jongwook** [comment]
The linked Keras example is not affiliated with us, and it seems that they used quite different architecture choices, such as using pre-trained Xception and BERT models, as opposed to training BiT/ResNet and autoregressive transformer from scratch.

Although we haven't experimented with their target matrix, I suspect that it'll be more difficult to train from scratch because the within-domain similarity is not as informative as the case of using pre-trained models; but I'm happy to be corrected on this.

**Comment by nishanthcgit** [comment]
Ok, I think I understand what you mean. If I do get to train and try it out, I will tell you about it! Thanks a lot.
